include/onnxruntime/core/platform/threadpool.h

```diff
@@ -46,6 +46,14 @@ class ThreadPool {
   */
   void ParallelFor(int32_t total, std::function<void(int32_t)> fn);
 
+  template <typename T>
+  static void ParallelFor(ThreadPool* tp, int32_t total, T&& fn) {
+    if (tp != nullptr)
```

Comment:
This seems like this would hide bad code. If there's no threadpool, does it make sense to be calling a function multiple times sequentially in an method called 'ParallelFor'?

I would have thought it better for the calling code to explicitly handle a nullptr for threadpool and do its work directly.

Reply:
It's not hiding something, by design the pointer could be nullptr, because people want to disable multiple threading at some time.

Reply:
This seems like a utility function that all callers can use so that not everyone needs to worry about whether the pool is nullptr or not. Putting this in the threadpool cIass can be confusing for clients to know which ParallelFor they should use. Should we call it TryParallelFor and place it in the concurrency namespace instead?

Reply:
To me this hides that the work is going to be done sequentially. Naively I would assume there was some other way for the implementation of ParallelFor to do parallelism if tp was nullptr. 

I understand that by design the threadpool could be nullptr, but if it is nullptr the code that is attempting to use ParallelFor should not really be attempting to use ParallelFor at all should it? I would have thought it should be doing its work directly in that case.

Reply:
Renamed it to TryParallelFor.